Реализация REST API для управления фильмами и пользователями

mqrkel opened the pull request:
- Разработана архитектура проекта на Spring Boot.
- Добавлены основные сущности: `Film`, `User`, `ResponseFilmDto`, `RequestFilmDto`, `RequestFilmWithIdDto`, `ResponseUserDto`, `RequestUserDto`, `RequestUserWithIdDto`.
- Реализованы контроллеры (`FilmController`, `UserController`) для управления фильмами и пользователями.
- Добавлен сервисный слой (`FilmServiceImpl`, `UserServiceImpl`) с обработкой логики работы с фильмами и пользователями.
- Внедрены хранилища `FilmMemoryStorage` и `UserMemoryStorage` с базовыми CRUD-операциями.
- Реализована валидация входных данных с использованием `Hibernate Validator`.
- Добавлены кастомные исключения (`FilmNotFoundException`, `UserNotFoundException`, `FilmCreationException`, `UserCreationException`, `FilmUpdateException`, `UserUpdateException`).
- Настроено логирование через `SLF4J`.
- Поддержка сериализации/десериализации JSON с помощью `Jackson`.

mqrkel committed:
Initial commit: Реализация REST API для управления фильмами и пользователями

- Разработана архитектура проекта на Spring Boot.
- Добавлены основные сущности: `Film`, `User`, `ResponseFilmDto`, `RequestFilmDto`, `RequestFilmWithIdDto`, `ResponseUserDto`, `RequestUserDto`, `RequestUserWithIdDto`.
- Реализованы контроллеры (`FilmController`, `UserController`) для управления фильмами и пользователями.
- Добавлен сервисный слой (`FilmServiceImpl`, `UserServiceImpl`) с обработкой логики работы с фильмами и пользователями.
- Внедрены хранилища `FilmMemoryStorage` и `UserMemoryStorage` с базовыми CRUD-операциями.
- Реализована валидация входных данных с использованием `Hibernate Validator`.
- Добавлены кастомные исключения (`FilmNotFoundException`, `UserNotFoundException`, `FilmCreationException`, `UserCreationException`, `FilmUpdateException`, `UserUpdateException`).
- Настроено логирование через `SLF4J`.
- Поддержка сериализации/десериализации JSON с помощью `Jackson`.

mqrkel committed:
Initial commit: Реализация REST API для управления фильмами и пользователями

- Разработана архитектура проекта на Spring Boot.
- Добавлены основные сущности: `Film`, `User`, `ResponseFilmDto`, `RequestFilmDto`, `RequestFilmWithIdDto`, `ResponseUserDto`, `RequestUserDto`, `RequestUserWithIdDto`.
- Реализованы контроллеры (`FilmController`, `UserController`) для управления фильмами и пользователями.
- Добавлен сервисный слой (`FilmServiceImpl`, `UserServiceImpl`) с обработкой логики работы с фильмами и пользователями.
- Внедрены хранилища `FilmMemoryStorage` и `UserMemoryStorage` с базовыми CRUD-операциями.
- Реализована валидация входных данных с использованием `Hibernate Validator`.
- Добавлены кастомные исключения (`FilmNotFoundException`, `UserNotFoundException`, `FilmCreationException`, `UserCreationException`, `FilmUpdateException`, `UserUpdateException`).
- Настроено логирование через `SLF4J`.
- Поддержка сериализации/десериализации JSON с помощью `Jackson`.

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/exception/FilmNotFoundException.java:
```diff
@@ -0,0 +1,7 @@
+package ru.yandex.practicum.filmorate.exception;
+
+public class FilmNotFoundException extends RuntimeException {
```
Лучше создать один класс исключение NotFoundException и в описании ошибки уже писать с чем связано исключение с user, film и тд..  в больших проектах может быть сотни  сущностей, зачем на каждую из них создавать отдельное исключение. Лучше сделать по одному исключению на каждый тип ошибки, а не сущности.

Аналогично и для остальных исключений.

mqrkel replied:
Поправил

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmServiceImpl.java:
```diff
@@ -0,0 +1,96 @@
+                .toList();
+    }
+
+    private ResponseFilmDto convertToResponseFilmDto(Film updatedFilm) {
```
private ResponseFilmDto convertToResponseFilmDto(Film updatedFilm) {
    private Film converToFilm(RequestFilmDto filmDto) {
.....

Такого рода преобразования (из одного типа объекта в другой) принято выделять в отдельные утилитарные классы, так называемые мапперы, например, FilmMapper и т.п.
Тоже и для пользователей.

mqrkel replied:
Поправил

avfyodorov commented on src/main/java/ru/yandex/practicum/filmorate/validator/ReleaseDate.java:
```diff
@@ -0,0 +1,18 @@
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.FIELD)
+@Constraint(validatedBy = ReleaseDateValidator.class)
+public @interface ReleaseDate {
```
Очень здорово, что Вы освоили создание собственных аннотаций. Теперь можно было бы сделать следующий шаг и определять дату как входной параметр:

ReleaseDate("1895-12-28")

Необязательно, на Ваше усмотрение.

mqrkel replied:
Добавил

avfyodorov reviewed (commented):
Добрый день, Максим! 

Хорошая работа, от меня всего пара уточнений. 
Да, ещё один момент, по техническому заданию нужны тесты, посмотрите, пожалуйста 
.------
Добавьте unit-тесты для валидации моделей. Убедитесь, что она работает на граничных условиях.

mqrkel committed:
Добавлены интеграционные тесты для контроллеров FilmController и UserController.
Рефакторинг исключений и преобразования объектов:
- Созданы универсальные исключения.
- Переработаны методы преобразования объектов `convertToResponseFilmDto`, `convertToFilm`,`convertToUser`,`convertToResponseUserDto` в отдельные утилитарные классы.
- Реализована возможность использования дат в аннотациях, `ReleaseDate("1895-12-28")`.

mqrkel committed:
Добавлены интеграционные тесты для контроллеров FilmController и UserController.
Рефакторинг исключений и преобразования объектов:
- Созданы универсальные исключения.
- Переработаны методы преобразования объектов `convertToResponseFilmDto`, `convertToFilm`,`convertToUser`,`convertToResponseUserDto` в отдельные утилитарные классы.
- Реализована возможность использования дат в аннотациях, `ReleaseDate("1895-12-28")`.

avfyodorov reviewed (commented):
Добрый вечер, Максим!

Всё в порядке, замечаний нет. 
Работа принята.
